### PR TITLE
Monoid wrappers for additive and multiplicative structures

### PR DESCRIPTION
**Pull request by jcranch**
These wrappers allow monoid functionality to be used on additive/multiplicative structures.

They also allow "sum" and "product" to work using foldMap rather than foldl. For some data structures, this may be more efficient.

### CODE REVIEW

**Comment by tonyday567** [comment]
Thanks so much for the PR.

I'm thinking about it.  It's technically the right way to go, but, in supplying a new Sum and Product, the library, with a purported aim of being a drop in replacement for Num and friends, would be reducing the footprint of these wrappers by not supplying the instances that Data.Semigroup does.  Ie the ones I list below - it's a lot of work.

We could call them Add and Mult to get around this, or we could ignore it and stomp on the more general usage that these instances represent, which I bet aren't used very often.

MonadFix Product	
 MonadZip Product	
 Foldable Product	
 Foldable1 Product	 
 Traversable Product	
 Applicative Product	
 Functor Product	
 Monad Product	
 Generic1 Product	 
 Data a => Data (Product a)	
 Bounded a => Bounded (Product a)	
 Generic (Product a)	 
 Num a => Num (Product a)	
 Read a => Read (Product a)	
 Show a => Show (Product a)	
 Eq a => Eq (Product a)	
 Ord a => Ord (Product a)	
 type Rep1 Product	
 type Rep (Product a)

**Comment by jcranch** [comment]
It's a solid point. I'd like to vote to keep them called Sum and Product. Add and Mult are just as good as names, but users have probably come to expect the existing names, and they literally do the same job.

I think this is actually not a huge amount of work, and I'm happy to have a go at it. I think I can classify these instances into several broad categories, although one could argue about precisely which lie in which:

1. Classes with build-in derivability, like `Read`, `Eq` and `Ord`.

2. `Functor`, which can be done with no boilerplate using the (safe, commonly used) language extension `DeriveFunctor`

3. Instances which don't really do anything (like `Foldable` and `Monad`).

4. Instances which are morally dubious (like `Num`; if you ask a mathematician whether the multiplicative monoid of a ring is a ring, they'll say no). However, I suppose they're harmless. If you want these, they can be done with the (safe, commonly used) language extension `GeneralizedNewtypeDeriving`.

**Comment by tonyday567** [comment]
OK, I like the categorization.  Num etc is banned, so 4 is out.

So if we get categories 1 and 2 in for sure.  I'd be happy to let category 3 slide - it was a trend to spam instances a while back but I think we're over that now.

As a reference:

```
instance Num a => Semigroup (Sum a) where
        (<>) = coerce ((+) :: a -> a -> a)
        stimes n (Sum a) = Sum (fromIntegral n * a)
```

Just wondering if a <> b = a + b is the same as <> = coerce +. I would be sad if coerce is better.

Also stimes should be included as we have our own Integral.

**Comment by jcranch** [comment]
It sounds like our opinions are converging.

Yes, I think that instances in class 3 are a bad thing, particularly because Haddock doesn't document instance definitions and so anyone sitting there thinking "how on earth is `Sum` a monad?" has no recourse except to reading the source code.

Since I wrote, I decided to relent a bit about class 4. Not `Num`, for the reasons you state, but I think one should probably use `GeneralizedNewtypeDeriving` to provide instances of your numeric class hierarchy. Even though the multiplicative monoid is not a ring, the ring operators retain their clear interpretation, and it is nicer to be able to do x + y than Product (getProduct x + getProduct y).

I'd be sad if coerce was better too! The ghc people are proud of their optimiser, and this optimisation is supposed to be trivial. I guess one could check the generated code.

File: src/NumHask/Algebra/Additive.hs [hunk 1]
```diff
@@ -46,12 +48,23 @@ class Additive a where
...
+instance Additive a => P.Monoid (Sum a) where
+  mempty = Sum zero
+
 -- | Compute the sum of a 'Data.Foldable.Foldable'.
 --
 -- >>> sum [0..10]
 -- 55
 sum :: (Additive a, P.Foldable f) => f a -> a
-sum = foldl' (+) zero
+sum = getSum P.. P.foldMap Sum
```
Comment (tonyday567) [inline src/NumHask/Algebra/Additive.hs:67]: Why not the left fold, foldMap'?

File: src/NumHask/Algebra/Additive.hs [hunk 2]
```diff
@@ -46,12 +48,23 @@ class Additive a where
 
   zero :: a
 
+-- | A wrapper for an Additive which distinguishes the additive structure
+newtype Sum a = Sum {
+  getSum :: a
+} deriving (Eq, Ord, Show)
```
Comment (tonyday567) [inline src/NumHask/Algebra/Additive.hs:54]: Read

Comment (tonyday567) [reply src/NumHask/Algebra/Additive.hs:54]: Generic, Generic1, Data with derivegeneric and derivedatatypeable

**Comment by tonyday567** [comment]
The other data/newtype declarations in the library have these instances:

Ratio: Show, Eq
Polar: Eq, Show, Generic
Complex: Eq, Show, Read, Data, Generic, Generic1, Functor, Foldable, Traversable

So a fairly random scattering.

To refine the hierarchy a bit (I'm practicing i before e!):

Eq, Show, Read, Ord

Generic, Generic1, Data are all generic programming constructs that seem benign even if somewhat mysterious. Include

Rep, Rep1 look like magic. avoid I suppose

Foldable, Traversable and friends seem a bit boiler-plate and dangerous to supply without a good reason. Remove from Complex. 

In terms of the internal numhask hierarchical shennanigans:

- You can't go backwards and derive a Ring instance for Sum cause Ring.hs depends on Additive.

- You can say this in Multiplicative:
```
instance (Multiplicative a) => Multiplicative (Sum a) where
  (*) (Sum a) (Sum b) = Sum (a * b)
  one = Sum one
```
but because you've imported NumHask.Algebra.Additive, you cant then do the same for Product as you get a cyclical dependency. If you really wanted it you'd have to put Sum and Product in separate modules - quite a horrible tradeoff.


- you can import NumHask.Algebra.Additive and then say 

```
newtype Sum a = Sum {
  getSum :: a
} deriving (Eq, Ord, Show, Read, Generic, Generic1, Multiplicative)
```
with generalised newtype deriving but then you can't do vice versa ie derive Additive for Product, because that gets cyclical again.

Which is a shame because I really, really like the explicit Multiplicative instance for Sum approach. `Sum one` certainly evokes the peano numbers. And `Product zero` could be called `the Annihilator` without loss of technical correctness.

FromInteger & FromRational

With these instances you could say getSum $ 1 <> 2 <> 3 because the type of 1 <> 2 <> 3 would be (FromInteger a, Monoid a)

**Comment by tonyday567** [comment]
Thanks for you interest in the library!

**Pull request merged**